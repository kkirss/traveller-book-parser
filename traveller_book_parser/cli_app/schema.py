import json
import logging
import pathlib
from typing import Optional

from pydantic import BaseModel
import typer

from traveller_book_parser.books.book_description import BookDescription
from traveller_book_parser.settings import SETTINGS
from traveller_book_parser.traveller_models.trav_object import TravObjectRoot
from traveller_book_parser.traveller_models.traveller_database import TravellerDatabase
from traveller_book_parser.utils import ensure_folder

logger = logging.getLogger(__name__)

schema_app = typer.Typer()


def dump_model_schema(
    path: Optional[pathlib.Path],
    cls: type[BaseModel],
    name: str,
):
    """Dump the JSON schema of a model."""
    if not cls.__doc__:
        logger.warning("Model '%s' has no docstring.", cls.__name__)

    if path is None:
        path = SETTINGS.default_schema_output_path / f"{name}.json"
    elif path.is_dir():
        path = path / f"{name}.json"

    ensure_folder(path.parent)

    with path.open("w") as f:
        json.dump(cls.model_json_schema(), f, indent=2)


def book_description_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump the JSON schema of the 'BookDescription' model."""
    dump_model_schema(path, BookDescription, "BookDescription")


schema_app.command("book_description")(book_description_schema_cli)
schema_app.command("BookDescription")(book_description_schema_cli)


def database_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump the JSON schema of the 'TravellerDatabase' model."""
    dump_model_schema(path, TravellerDatabase, "TravellerDatabase")


schema_app.command("traveller_database")(database_schema_cli)
schema_app.command("TravellerDatabase")(database_schema_cli)


def trav_object_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump the JSON schema of the 'TravObject' model."""
    dump_model_schema(path, TravObjectRoot, "TravObject")


schema_app.command("trav_object")(trav_object_schema_cli)
schema_app.command("TravObject")(trav_object_schema_cli)


def all_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump all JSON schema files."""
    book_description_schema_cli(path)
    database_schema_cli(path)
    trav_object_schema_cli(path)


schema_app.command("all")(all_schema_cli)
