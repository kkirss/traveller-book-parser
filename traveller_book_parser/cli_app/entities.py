from traveller_book_parser.traveller_models.entity import Entity


def print_entities(entities: list[Entity]):
    for entity in entities:
        print(entity)
