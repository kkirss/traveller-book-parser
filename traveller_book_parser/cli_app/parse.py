from traveller_book_parser.books.parse_books import parse_all_books, parse_book
from traveller_book_parser.traveller_database.output import outputted_database
from traveller_book_parser.traveller_database.output_format.output_format import (
    DatabaseOutputFormat,
)


def parse_book_cli(
    book_code_name: str,
    output_format: DatabaseOutputFormat = DatabaseOutputFormat.json,
):
    """Parse a book and print processed entities."""
    with outputted_database(output_format) as database:
        parse_book(database, book_code_name)


def parse_all_books_cli(
    output_format: DatabaseOutputFormat = DatabaseOutputFormat.json,
):
    """Parse all books and print processed entities."""
    with outputted_database(output_format) as database:
        parse_all_books(database)
