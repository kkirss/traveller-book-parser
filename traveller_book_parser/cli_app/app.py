import typer

from .parse import parse_all_books_cli, parse_book_cli

app = typer.Typer(
    help="Parse Traveller books into other formats.",
    no_args_is_help=True,
)


app.command("parse-book", no_args_is_help=True)(parse_book_cli)
app.command("parse")(parse_all_books_cli)


def run_cli():
    """Run CLI."""
    app()
