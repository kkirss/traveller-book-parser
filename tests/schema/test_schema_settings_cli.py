from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion
from typer.testing import CliRunner

from traveller_book_parser.cli_app.app import app


@pytest.mark.parametrize(
    ("command_name", "check_snapshot"),
    [
        ("Settings", True),
        ("settings", False),
    ],
)
def test_schema_settings_cli(
    *,
    command_name: str,
    check_snapshot: bool,
    cli_runner: CliRunner,
    snapshot: SnapshotAssertion,
    tmp_path: Path,
):
    # Arrange
    command = ["schema", command_name]
    output_file = tmp_path / f"{'.'.join(command)}.json"

    # Act
    result = cli_runner.invoke(app, [*command, "--path", str(output_file)])

    # Assert
    assert result.exit_code == 0, result.output
    assert output_file.exists()

    with output_file.open("r", encoding="utf-8") as f:
        schema = f.read()

    if check_snapshot:
        # Compare the schema with the snapshot
        assert schema == snapshot
