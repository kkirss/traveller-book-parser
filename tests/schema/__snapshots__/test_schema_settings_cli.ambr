# serializer version: 1
# name: test_schema_settings_cli[Settings-True]
  dict({
    '$defs': dict({
      'LogLevel': dict({
        'description': 'Logging level.',
        'enum': list([
          'CRITICAL',
          'ERROR',
          'WARNING',
          'INFO',
          'DEBUG',
          'NOTSET',
        ]),
        'title': 'LogLevel',
        'type': 'string',
      }),
    }),
    'description': 'Settings for the application.',
    'properties': dict({
      'book_descriptions_path': dict({
        'description': 'Folder path where book description files are loaded from',
        'format': 'path',
        'title': 'Book Descriptions Path',
        'type': 'string',
      }),
      'books_path': dict({
        'description': 'Folder path where books are loaded from',
        'format': 'path',
        'title': 'Books Path',
        'type': 'string',
      }),
      'cache_path': dict({
        'description': 'Path for temporary files',
        'format': 'path',
        'title': 'Cache Path',
        'type': 'string',
      }),
      'default_output_path': dict({
        'description': 'Path for output files',
        'format': 'path',
        'title': 'Default Output Path',
        'type': 'string',
      }),
      'default_schema_output_path': dict({
        'description': 'Path for schema output files',
        'format': 'path',
        'title': 'Default Schema Output Path',
        'type': 'string',
      }),
      'images_path': dict({
        'description': 'Path for image files',
        'format': 'path',
        'title': 'Images Path',
        'type': 'string',
      }),
      'log_intermediate_data': dict({
        'default': False,
        'description': 'Whether to log intermediate data. Note: These are logged at DEBUG level.',
        'title': 'Log Intermediate Data',
        'type': 'boolean',
      }),
      'log_level': dict({
        'allOf': list([
          dict({
            '$ref': '#/$defs/LogLevel',
          }),
        ]),
        'default': 'INFO',
      }),
      'log_parsed_objects': dict({
        'default': False,
        'description': 'Whether to log parsed objects. Note: These are logged at DEBUG level.',
        'title': 'Log Parsed Objects',
        'type': 'boolean',
      }),
      'pdf_to_html_executable': dict({
        'description': 'Path to pdftohml (version 4.x) executable',
        'format': 'path',
        'title': 'Pdf To Html Executable',
        'type': 'string',
      }),
    }),
    'title': 'Settings',
    'type': 'object',
  })
# ---
